### PR TITLE
Fix NPE caused by AbstractScriptFileWatcher initialization order

### PR DESCRIPTION
**Pull request by jimtng**
See the discussion in #3445

Another problem I encountered was the onReadyMarkerAdded got called out of order. It jumped from 80 -> 20 -> 100 -> 40 -> 50 -> 70 -> 30. Here's an example trace:
```
023-03-13 01:42:22.428 [TRACE] [ort.loader.AbstractScriptFileWatcher] - onReadyMarkerAdded(startlevel=80) 100 => 100 for /openhab/conf/automation/ruby
2023-03-13 01:42:22.428 [TRACE] [ort.loader.AbstractScriptFileWatcher] - onReadyMarkerAdded(startlevel=20) 100 => 100 for /openhab/conf/automation/ruby
2023-03-13 01:42:22.429 [TRACE] [ort.loader.AbstractScriptFileWatcher] - onReadyMarkerAdded(startlevel=100) 100 => 100 for /openhab/conf/automation/ruby
2023-03-13 01:42:22.430 [TRACE] [ort.loader.AbstractScriptFileWatcher] - onReadyMarkerAdded(startlevel=40) 100 => 100 for /openhab/conf/automation/ruby
2023-03-13 01:42:22.430 [TRACE] [ort.loader.AbstractScriptFileWatcher] - onReadyMarkerAdded(startlevel=50) 100 => 100 for /openhab/conf/automation/ruby
2023-03-13 01:42:22.431 [TRACE] [ort.loader.AbstractScriptFileWatcher] - onReadyMarkerAdded(startlevel=70) 100 => 100 for /openhab/conf/automation/ruby
2023-03-13 01:42:22.431 [TRACE] [ort.loader.AbstractScriptFileWatcher] - onReadyMarkerAdded(startlevel=30) 100 => 100 for /openhab/conf/automation/ruby
```

This caused AbstractScriptFileWatcher to stop loading the files because the last startLevel it received through onReadyMarkerAdded() was startLevel 30, so it never end up instructing importFileWhenReady to run.

### CODE REVIEW

**Comment by jimtng** [comment]
@J-N-K I'd appreciate your review on this

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java [hunk 1]
```diff
@@ -272,7 +277,7 @@ private CompletableFuture<Void> removeFile(String scriptIdentifier) {
                 ScriptFileReference ref = scriptMap.remove(scriptIdentifier);
 
                 if (ref == null) {
-                    logger.warn("Failed to unload script '{}': script reference not found.", scriptIdentifier);
+                    logger.debug("Failed to unload script '{}': script reference not found.", scriptIdentifier);
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:280]: Why did you reduce this logger? If unloading fails it is very likely we create memory or thread leaks.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java [hunk 2]
```diff
@@ -353,18 +358,14 @@ private boolean createAndLoad(ScriptFileReference ref) {
         return false;
     }
 
-    private void initialImport() {
-        File directory = watchPath.toFile();
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:357]: Why did you remove the directory creation?

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java [hunk 3]
```diff
@@ -353,18 +358,14 @@ private boolean createAndLoad(ScriptFileReference ref) {
...
-
-        if (!directory.exists()) {
-            if (!directory.mkdirs()) {
-                logger.warn("Failed to create watched directory: {}", watchPath);
-            }
-        } else if (directory.isFile()) {
-            logger.warn("Trying to watch directory {}, however it is a file", watchPath);
+    protected void performInitialImport() {
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:361]: I don't think this should be a separate method. If a sub-class wants to handle the import one it's own, it's free to do so and nesting methods only adds complexity without a real benefit.

**Review by J-N-K — CHANGES_REQUESTED** [review]
I think we should better investigate why the ready markers arrive unordered. Or use the ready marker as trigger only to get the actual start level from the `StartLevelService`.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java (hunk 1 above)
Comment (jimtng) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:280]: The problem is that this would happen when some random, untracked files got deleted.
Example:
- /conf/automation/ruby/ is the main watchPath
- /conf/automation/ruby/a_file is excluded and not added to scriptMap (for whatever reason - e.g. it is excluded by AbstractScriptFileWatcher itself, or excluded by child class)
- that file got deleted, triggering processWatchEvent
- Now people see this warning

It happens consistently with Ruby because .gem subdir is excluded and during addon initialisation, some lock files were created and then deleted, triggering a watchEvent.

Secondly - I couldn't see a memory / thread leak risk here.

In all places that call `removeFile`, they checked whether the entry they're removing exists inside scriptMap first before calling it. One place this didn't happen is inside processWatchEvent for DELETE event - and that's exactly when this happened.

Would you prefer that a check is done in processWatchEvent too, and leave the WARN logging there? I'd be happy with that solution as well, although I think it's redundant. Are there any other reasons that scriptMap.remove would return null other than the fact that the given element didn't exist?

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java (hunk 2 above)
Comment (jimtng) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:357]: I thought it's redundant because:
- The directory creation is also done in WatchServiceImpl
- The directory creation is also done in `activate()` - this is also redundant here, but happy to leave it there to ensure this behaviour in case it changed in WatchServiceImpl.
- Prior to this PR, the directory creation is checked/performed inside `initialImport`

There is only a tiny window between activate() and the time the final call to `initialImport()` was made upon reaching startlevel 100 that an existing directory ensured by `activate()` got deleted. However, I'm happy to restore it if you think it's essential.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java (hunk 3 above)
Comment (jimtng) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:361]: The only reason I made it a separate method, is so that there was this check inside the constructor:
```
        if (currentStartLevel > StartLevelService.STARTLEVEL_MODEL) {
```
and I don't want the child class to need to know this implementation detail, and instead it should [just call `performInitialImport()`](https://github.com/openhab/openhab-addons/pull/14590/files#diff-0391e0939589717a418a22a3138a32b9fe3b36a31091bb5bfbb8cd5bd9938101R66).

I couldn't move it into `initialImport` because it's not always checked with this exact condition everywhere, just inside the constructor.

**Comment by jimtng** [comment]
> I think we should better investigate why the ready markers arrive unordered.

I agree. However I just wanted my jruby addon to load so I can continue using 4.0-snapshot at this point. We could dig into the root cause of the out-of-order start level events later and remove the check here once we've fixed that.



> Or use the ready marker as trigger only to get the actual start level from the `StartLevelService`

I will try this.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java [hunk 4]
```diff
@@ -225,25 +230,25 @@ private List<Path> listFiles(Path path, boolean includeSubDirectory) {
...
+                String scriptIdentifier = ScriptFileReference.getScriptIdentifier(fullPath);
+                if (scriptMap.containsKey(scriptIdentifier)) {
+                    removeFile(scriptIdentifier);
                 }
             }
-
-            if (file.canRead() && (kind == CREATE || kind == MODIFY)) {
-                addFiles(listFiles(file.toPath(), watchSubDirectories));
-            }
+        } else if (!file.isHidden() && file.canRead() && (kind == CREATE || kind == MODIFY)) {
```
Comment (jimtng) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:250]: I moved the hidden file check here, because they should be checked during deletion (above), in case they got added previously. Otherwise they'd never be deleted.

**Comment by jimtng** [comment]
> think we should better investigate why the ready markers arrive unordered. Or use the ready marker as trigger only to get the actual start level from the `StartLevelService`.

I've investigated this further. 

Ready markers do arrive in order - IF you registered your tracker early on, before the actual startlevels were reached. However, if you registered your tracker very late, say when startlevel was already at 80, ReadyServiceImpl::registerTracker() will catch you up on it, and send all the already reached startlevels all in one go. This is also why previously the import seems to have gotten done twice.

Because ReadyService doesn't know that it's a "startlevel" and supposed to be sorted, it just sends it unordered. This is the first time I looked at ReadyService, but I presumed there could be other Ready marker stuff that makes no sense to be sorted? If it needs to be sorted, it can be done in a separate PR. It doesn't affect this PR.

This PR just makes sure that it doesn't execute the code when it's out of order. Also checking that the import hasn't been done before calling initialImport stops us from doing initialimport twice here.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java (hunk 3 above)
Comment (jimtng) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:361]: I realised that the two checks are essentially the same, and I've combined the two functions into one and use the same check.

**Comment by J-N-K** [comment]
> Ready markers do arrive in order - IF you registered your tracker early on, before the actual startlevels were reached. However, if you registered your tracker very late, say when startlevel was already at 80, ReadyServiceImpl::registerTracker() will catch you up on it, and send all the already reached startlevels all in one go. This is also why previously the import seems to have gotten done twice.
> 
> Because ReadyService doesn't know that it's a "startlevel" and supposed to be sorted, it just sends it unordered. This is the first time I looked at ReadyService, but I presumed there could be other Ready marker stuff that makes no sense to be sorted? If it needs to be sorted, it can be done in a separate PR. It doesn't affect this PR.

I checked and will provide a fix for that. The set that stores the received ready markers should retain the order (which it currently does not).

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java (hunk 2 above)
Comment (jimtng) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:357]: I've tested deleting conf/automation/jsr223 then starting openhab up. It got recreated, and it monitors changes in the directory. I tried creating a script in there and it got (re)loaded. So I believe it's fine to remove it from initialImport.

**Comment by jimtng** [comment]
Sorry for the late force-pushes. I believe it should be ready now.

> > Ready markers do arrive in order - IF you registered your tracker early on, before the actual startlevels were reached. However, if you registered your tracker very late, say when startlevel was already at 80, ReadyServiceImpl::registerTracker() will catch you up on it, and send all the already reached startlevels all in one go. This is also why previously the import seems to have gotten done twice.
> > Because ReadyService doesn't know that it's a "startlevel" and supposed to be sorted, it just sends it unordered. This is the first time I looked at ReadyService, but I presumed there could be other Ready marker stuff that makes no sense to be sorted? If it needs to be sorted, it can be done in a separate PR. It doesn't affect this PR.
> 
> I checked and will provide a fix for that. The set that stores the received ready markers should retain the order (which it currently does not).

Awesome, thanks! I don't think it makes a difference to this PR though. The checks here should still remain.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java [hunk 5]
```diff
@@ -225,25 +230,25 @@ private List<Path> listFiles(Path path, boolean includeSubDirectory) {
...
-                    if (watchSubDirectories) {
-                        synchronized (this) {
-                            String prefix = fullPath.getParent().toString();
-                            Set<String> toRemove = scriptMap.keySet().stream().filter(ref -> ref.startsWith(prefix))
-                                    .collect(Collectors.toSet());
-                            toRemove.forEach(this::removeFile);
-                        }
+
+        if (kind == DELETE) {
+            if (file.isDirectory()) {
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:235]: I wonder if this check is necessary. Did you check if we receive events for each file in the deleted directory? In that case we could ignore directory changes at all.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java [hunk 6]
```diff
@@ -144,6 +146,9 @@ public void activate() {
         } else if (!Files.isDirectory(watchPath)) {
             logger.warn("Trying to watch directory {}, however it is a file", watchPath);
         }
+
+        manager.addFactoryChangeListener(this);
+        readyService.registerTracker(this, new ReadyMarkerFilter().withType(StartLevelService.STARTLEVEL_MARKER_TYPE));
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:151]: Why did you move these?

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java (hunk 3 above)
Comment (J-N-K) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:361]: Thanks, can you keep the original name?

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java [hunk 7]
```diff
@@ -378,16 +375,25 @@ public void onDependencyChange(String scriptIdentifier) {
 
     @Override
     public synchronized void onReadyMarkerAdded(ReadyMarker readyMarker) {
+        int newLevel = Integer.parseInt(readyMarker.getIdentifier());
+
+        if (newLevel <= currentStartLevel) {
+            logger.trace("onReadyMarkerAdded({}) - ignored {}", readyMarker, watchPath);
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:381]:
```suggestion
            logger.trace("Ignored ready marker {} because it is for a lower start-level", readyMarker);
```

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java [hunk 8]
```diff
@@ -378,16 +375,25 @@ public void onDependencyChange(String scriptIdentifier) {
...
         if (currentStartLevel < StartLevelService.STARTLEVEL_STATES) {
             // ignore start level less than 30
             return;
         }
 
-        if (currentStartLevel == StartLevelService.STARTLEVEL_STATES) {
-            initialImport();
+        if (!ifInitialized().isDone()) {
+            performInitialImport();
         } else {
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:397]: Nice

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java (hunk 1 above)
Comment (jimtng) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java]: Another update on this. Although adding the check to make sure that the given scriptIdentifier exists in scriptMap inside processWatchEvent DELETE does indeed avoid triggering that warning, I came across another case where that warning was erroneously issued despite it being perfectly fine:

During bundle termination (e.g. bundle:stop), removeFile was called twice: once by `factoryRemoved` and another time by `deactivate`. And because removeFile is async, the same scriptIdentifier gets included twice in the removal. The first one removed the file normally, unloading the scriptEngine and all. The second one would find that the scriptIdentifier no longer exists in scriptMap and then issued this warning.

Instead, I've changed the warning to be issued when the scriptIdentifier is still there when it shouldn't be.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java (hunk 3 above)
Comment (jimtng) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java]: I've changed it to the original name.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java (hunk 5 above)
Comment (jimtng) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java]: You are absolutely right! I did some tests and deleting a directory with scripts inside it would correctly issue a DELETE event for each file inside that directory, but not the directory itself. This is because it's filtered by WatchServiceImpl so processWatchEvent would never receive events about directories anyway, so yes this piece of code will never be reached.

I have removed it and also adjusted the corresponding test.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java (hunk 6 above)
Comment (jimtng) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:151]: Remember that readyService.registerTracker will catch you up on all the startlevel markers that had already been received/reached? Well, that would trigger initialImport right there if I left it in the constructor.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java (hunk 7 above)
Comment (jimtng) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java]: I've changed it as suggested. However, I added the watchPath because when you have multiple instances of this class, you can't tell which is which otherwise.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java [hunk 9]
```diff
@@ -378,16 +375,25 @@ public void onDependencyChange(String scriptIdentifier) {
...
+        if (newLevel <= currentStartLevel) {
+            logger.trace("onReadyMarkerAdded({}) - ignored {}", readyMarker, watchPath);
+            return;
+        }
+
         int previousLevel = currentStartLevel;
-        currentStartLevel = Integer.parseInt(readyMarker.getIdentifier());
+        currentStartLevel = newLevel;
+
+        logger.trace("onReadyMarkerAdded({}) {} => {} {}", readyMarker, previousLevel, currentStartLevel, watchPath);
```
[the remark this answers is not in the document]
Comment (jimtng) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java]: I've changed it as suggested. However, I added the watchPath because when you have multiple instances of this class, you can't tell which is which otherwise.

**Comment by jimtng** [comment]
I've just realised, that we could avoid having two constructors by simply moving `initialImport` off the constructor and into `activate()`.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java [hunk 10]
```diff
@@ -378,15 +358,26 @@ public void onDependencyChange(String scriptIdentifier) {
...
+        logger.trace("Added ready marker {}: start level changed from {} to {}. watchPath: {}", readyMarker,
+                previousLevel, currentStartLevel, watchPath);
 
         if (currentStartLevel < StartLevelService.STARTLEVEL_STATES) {
             // ignore start level less than 30
             return;
         }
 
-        if (currentStartLevel == StartLevelService.STARTLEVEL_STATES) {
+        if (!ifInitialized().isDone()) {
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:380]:
```suggestion
        if (initialized.isDone()) {
```

**Review by J-N-K — COMMENTED** [review]
Thanks, code looks good now. I'll test it in the next days and if I don't encounter problems we can merge it.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java (hunk 6 above)
Comment (jimtng) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java:142]: Indeed. At first I only wanted to avoid the NPE and didn't want to change the existing code much. But you're correct, there's no need to do initialImport twice. I've removed it.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java (hunk 10 above)
Comment (jimtng) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java]: changed as requested

**Comment by jimtng** [comment]
> Thanks, code looks good now. I'll test it in the next days and if I don't encounter problems we can merge it.

Thanks for your help!

**Review by J-N-K — APPROVED** [review]

**Comment by J-N-K** [comment]
Seems to work fine for me.

**Comment by florian-h05** [comment]
Big thanks for that fix @jimtng, the error also affected JS Scripting!
